### DISCUSSION TITLE
Some tmux features I'm missing

### DISCUSSION THREAD

**Discussion by skanev** [wezterm/wezterm#2778] [Q&A]
Hey!

I want to start (again) by saying how amazing WezTerm is and how much I love it, and a big THANK YOU to everybody involved it making it happen.

I've started playing with the idea of getting rid of tmux and replacing with WezTerm completely. I have quite an involved setup, as I'm using tmux as a programmable (or at least, controllable) interface to my terminal, rather than just to have tabs. There are few things that fit into wezterm quite nicely, but there are two features I'm missing. Specifically:

1. `pipe-pane`. In tmux, I can get all the input going into a pane and react on it. I use it for a bunch of things, including paying attention to test failures and storing them somewhere so I can either load them in vim or rerun only the failures. Yes, I know that there are simpler ways to achieve it, but I have found this one to be quite productive. I don't need to change the input (e.g. I don't need to rewrite it or anything), just to be able to pipe it to a program. Is there a way to do that? I've been looking for an event, but I couldn't find one.
2. `show-popup`. I occasionally run popups for different things. For example, I have a script that opens that finds all active tmux session and gives them to fzf. I can select one and switch the session. I've figured out that sessions map to the `workspace` concept in WezTerm, but what I'm not sure how to do is how I can open a temporary/scratch terminal (ideally ontop of everything else). I can spin off a separate tab (or a split), but I'm looking for a neater way to do it.

Thanks in advance!

**Comment by bew** [comment]
For `pipe-pane`, you could define an action callback using https://wezfurlong.org/wezterm/config/lua/wezterm/action_callback.html , then in the callback function, use https://wezfurlong.org/wezterm/config/lua/pane/get_lines_as_text.html on the pane object.
(I don't see a way to only get the last command's output, using semantic zones. @wez you know if it's planned to have Lua access to semantic zones?)

There is an example (a bit hard to find) of retrieving content of a pane and opening it in vim. You can find it in the docs of custom events (which is the mecanism used by `action_callback`): https://wezfurlong.org/wezterm/config/lua/wezterm/on.html#custom-events
But you should be able to do anything you want in your callback, including piping that output in some other command!

**Reply by skanev** [reply]
Thanks!

I've seen those before, but I actually want to listen to the content of the pane passively. That is, I don't want to have to hit a key (or trigger an event) and then go fetch lines (and figure out what I've already processed and what I haven't). I'd prefer to get the raw stdout of whatever the program is outputting (happy to deal with control characters and all that) and tee it into another program.

**Comment by cd-a** [comment]
Popups would be great, also the only thing missing for me.

**Reply by haphamdev** [reply]
I'm  also waiting for the popup.

In tmux, I wrote [a shell script](https://github.com/haphamdev/dot-files/blob/master/tmux/switch-tmux-window.fish) to list all tmux windows (or sessions),use `fzf` to quickly pick a window.
[Then I use tmux popup to run that script.](https://github.com/haphamdev/dot-files/blob/master/tmux/tmux.conf#L146C1-L146C73)

**Reply by haphamdev** [reply]
[InputSelector seems to be a good choice for my use case](https://wezfurlong.org/wezterm/config/lua/keyassignment/InputSelector.html?h=input)
